src/librustc_mir/borrow_check.rs

```diff
@@ -931,41 +978,107 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
-                        self.describe_lvalue(lvalue));
+                    let item_msg = match self.describe_lvalue(lvalue) {
+                        Some(name) => format!("`{}`", name),
+                        None => "item".to_owned()
+                    };
+                    span_bug!(span, "&unique borrow for {} should not fail", item_msg);
                 }
             },
             Write(WriteKind::MutableBorrow(BorrowKind::Mut)) => {
-                if let Err(lvalue_err) = self.is_mutable(lvalue) {
+                if let Err(lvalue_err) = self.is_mutable(lvalue, is_local_mutation_allowed) {
+                    error_reported = true;
+
+                    let item_msg = match self.describe_lvalue(lvalue) {
+                        Some(name) => format!("immutable item `{}`", name),
+                        None => "immutable item".to_owned()
+                    };
+
                     let mut err = self.tcx.cannot_borrow_path_as_mutable(span,
-                        &format!("immutable item `{}`",
-                                  self.describe_lvalue(lvalue)),
+                        &item_msg,
                         Origin::Mir);
                     err.span_label(span, "cannot borrow as mutable");
 
                     if lvalue != lvalue_err {
-                        err.note(&format!("Value not mutable causing this error: `{}`",
-                            self.describe_lvalue(lvalue_err)));
+                        if let Some(name) = self.describe_lvalue(lvalue_err) {
+                            err.note(&format!("Value not mutable causing this error: `{}`", name));
+                        }
+                    }
+
+                    err.emit();
+                }
+            },
+            Write(WriteKind::Mutate) => {
+                if let Err(lvalue_err) = self.is_mutable(lvalue, is_local_mutation_allowed) {
+                    error_reported = true;
+
+                    let item_msg = match self.describe_lvalue(lvalue) {
+                        Some(name) => format!("immutable item `{}`", name),
+                        None => "immutable item".to_owned()
+                    };
+
+                    let mut err = self.tcx.cannot_assign(span,
+                        &item_msg,
+                        Origin::Mir);
+                    err.span_label(span, "cannot mutate");
+
+                    if lvalue != lvalue_err {
+                        if let Some(name) = self.describe_lvalue(lvalue_err) {
+                            err.note(&format!("Value not mutable causing this error: `{}`", name));
+                        }
                     }
 
                     err.emit();
                 }
             },
-            _ => {}// Access authorized
+            Write(WriteKind::Move) |
+            Write(WriteKind::StorageDeadOrDrop) |
+            Write(WriteKind::MutableBorrow(BorrowKind::Shared)) => {
+                let item_msg = match self.describe_lvalue(lvalue) {
```

Comment:
1. I think you're missing a check for whether the access is actually permitted, similarly to the previous lines. i.e. an
```Rust
                if let Err(lvalue_err) = self.is_mutable(lvalue, is_local_mutation_allowed) { 
```

2. There's no reason to use pretty-printing in `delay_span_bug` - it can hide information and it just makes the code uglier. Just format the lvalue directly (`format!("Accessing {:?} with the kind `{:?}` shouldn't be possible", lvalue, kind)`).